numpy/core/numeric.py

```diff
@@ -374,13 +378,74 @@ def extend_all(module):
         if a not in adict:
             __all__.append(a)
 
-newaxis = None
 
+def count_nonzero(a, axis=None):
+    """
+    Counts the number of non-zero values in the array ``a``.
+
+    Parameters
+    ----------
+    a : array_like
+        The array for which to count non-zeros.
+
+    axis : int, optional
+        Axis along which to count non-zeros. Default is None,
+        meaning that non-zeros will be counted along a flattened
+        version of ``a``.
+
+        .. versionadded:: 1.12.0
+
+    Returns
+    -------
+    count : int or array of int
+        Number of non-zero values in the array along a given axis.
+        Otherwise, the total number of non-zero values in the array
+        is returned.
+
+    See Also
+    --------
+    nonzero : Return the coordinates of all the non-zero values.
+
+    Examples
+    --------
+    >>> np.count_nonzero(np.eye(4))
+    4
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]])
+    5
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]], axis=0)
+    array([1, 1, 1, 1, 1])
+    >>> np.count_nonzero([[0,1,7,0,0],[3,0,0,2,19]], axis=1)
+    array([2, 3])
+
+    """
+    from numpy import apply_along_axis, integer, issubdtype
+
+    a = asanyarray(a)
+
+    if (issubdtype(a.dtype, integer) or
+            issubdtype(a.dtype, float) or
+            issubdtype(a.dtype, complex)):
+        a = (a != 0)
+
+    elif issubdtype(a.dtype, str):
+        a = (a != '')
+
+    if a.dtype.name == 'bool':
+        if axis is None:
+            return multiarray.count_nonzero(a)
+
+        return a.sum(axis=axis, dtype=longlong)
+
+    if axis is None:
+        return multiarray.count_nonzero(a)
+
+    return apply_along_axis(multiarray.count_nonzero, axis, a)
```

Comment:
Because element-wise comparison does not do very well for complex `dtypes`, I'm going to leave this part of the code <b>as is</b> because I'll have checks to catch the "nicer" cases.